features/323-login-id/stage-1.md

```diff
@@ -0,0 +1,93 @@
+# Implict login ID key, Login ID realm
+
+## Database Schema
+
+- In `_auth_provider_password` table:
+    - A column `realm` (`TEXT NOT NULL`) would be added.
+    - The existing unique index on `login_id` would be changed to include `realm`.
+
+## Configuration
+
+- Add field `realmWhitelist` (list of strings, optional, default to `["default"]`).
+
+## HTTP API
+- `POST /login`:
+    - Change type of `login_id` from map of strings to a plain string value.
+    - Add parameter `realm` (string, optional).
+    - Example:
+      ```json
+        {
+            "login_id": "test@example.com",
+            "password": "12345678"
+        }
+        {
+            "realm": "student",
+            "login_id": "test@example.com",
+            "password": "12345678"
+        }
+      ```
+- `POST /signup`
+    - Add parameter `realm` (string, optional).
+    - Example:
+      ```json
+        {
+            "realm": "teacher",
+            "login_ids": {
+                "email": "test@example.com"
+            },
+            "password": "12345678"
+        }
+        {
+            "realm": "student",
+            "login_id": {
+                "email": "test@example.com"
+            },
+            "password": "12345678"
+        }
+      ```
+
+- `POST /verify_request`
+    - Rename parameter `record_key` to `login_id_key`
+    - Add parameter `login_id` (string, optional).
+    - Example: 
+      ```json
+        {
+            "login_id_key": "email",
+            "login_id":"test@example.com"
+        }
+      ```
+
+
+## SDK API
+Following snippets use TypeScript:
+```diff
+- async signup(loginIDs: { [key: string]: string }, password: string, data?: unknown): Promise<User>;
+- async signupWithUsername(username: string, password: string, data?: unknown): Promise<User>;
+- async signupWithEmail(email: string, password: string, data?: unknown): Promise<User>;
++ async signup(loginIDs: { [key: string]: string }, password: string, data: unknown, realm?: string): Promise<User>;
++ async signupWithUsername(username: string, password: string, data?: unknown, realm?: string): Promise<User>;
++ async signupWithEmail(email: string, password: string, data?: unknown, realm?: string): Promise<User>;
+
+- async login(loginID: { [key: string]: string }, password: string): Promise<User>;
+- async loginWithUsername(username: string, password: string): Promise<User>;
+- async loginWithEmail(email: string, password: string): Promise<User>;
++ async login(loginID: string, password: string, realm?: string): Promise<User>;
```

Comment:
I think we should still support login with key, (and maybe loginWithUsername and loginWithEmail too)

The reason is, as my understand part of it is for semantic meaning of loginID, it would be more convenient if the end user login with Phone No / Email, and if it is known to be phone no / email, we can do the normalization for user to match with login ID.

(e.g. normalize of country code + area code for phone number login, or postfix match if login by phone no, etc)

Reply:
Design of login ID normalization is not in the scope of this issue. Since login ID key is not used during the login process in this moment, I'll prefer to add login ID key if needed later.

Reply:
So what is the scope of this issue actually...

The objective states above are:

```
1. Allow auth related features to understand semantics of some login IDs
2. Support multiple login ID with same login ID key
3. Support compound login IDs
4. Support adding/removing login ID
```

I started to feel confused with what is this about and why we are changing many things now.

The only reason for 1 and 2 to exists, to my understanding, is to allow normalization of login ID in the case of email addresses and phone number. Otherwise we have no reason to make auth aware of the semantic of some Login IDs; and hence no resaon why we need to support multiple login ID with same login ID key.

Support compound login IDs is good, but basically it is only adding realm.

Support adding/removing login ID was well covered in the current spec, just not implemented yet.

Can we clarify what we are trying to design here and move to implementation and delivery asap?

Reply:
Point 1 is also for user verification & forgot password: they require understanding of login ID key so that they can send messages to user.
Point 2 is to address use case 6 as mentioned in design document: allow use case such as Facebook, where user can login with any verified email associated with the user.
I'll remove point 4 for future discussion on principal APIs.

Reply:
@kiootic Actually in the current design it was not needed:

1. User verification / forgot password doesn't require understanding of login ID key to send message, as long as the login ID key's type (email, or phone no) is specified in the functions, e.g. in #294 

`requestVerification(LoginIDKey: String, type: ["email", "phone"])`

2. Use case 6 doesn't need pt 2 neither, it was supposed to be done in something like these in #294. (As login ID key could be arbitrary. 

```
login_ids: {
      email: 'example@example.com'
      email2: 'example2@example.com'
    },

```

Reply:
Please do not remove #4, and supplement how that could be done instead (you can even just copy and paste the API proposed in #294 .

I don't think we should have a new specification which provide less features than the proposal it deprecated.

Especially as it would change our APIs / SDK interfaces back-and-forth.

Reply:
- I'm strongly against that the type of login ID key is determined by client side; we should not allow too much control by client side. The same goes for arbitrary login ID key; I'd prefer a sensible default white-list of keys rather than unrestricted login ID keys.
- I'll add back create/delete login ID API, however update login ID API has to go; it makes the identity of principals mutable and is dangerous from a security point of view.

Reply:
> * I'm strongly against that the type of login ID key is determined by client side; we should not allow too much control by client side. The same goes for arbitrary login ID key; I'd prefer a sensible default white-list of keys rather than unrestricted login ID keys.

I think in this design the Login ID Key is whitelist by a server-side configuration only so that's a non-issue? (In fact at #294 there is a option to whitelist it already)



>  * I'll add back create/delete login ID API, however update login ID API has to go; it makes the identity of principals mutable and is dangerous from a security point of view.

Can you be specific on why `updateLoginID` would be dangerous even for Client side with a logged in user session? Beware it is an API to update the LoginID, not other data of the principals.

But I'm fine with ignoring the updateLoginID one, coz technically it is equivalent to create and remove (see the API changes below).

BTW, when you add back create/delete login ID, please beware that in #294 , it is expected that each login ID key would have ONE login ID value; In this spec, you make each login ID key could have multiple Login ID value, so although we can have the same function name of `create/update/removeLoginID`, the parameter probably need to be changed.

Maybe something like this:

```
createLoginID(loginIDKey: String, loginIDValue: String) // this will add a new login ID to the existing array of login ID value if it already exists 
updateLoginID(loginIDKey: String, originalLoginIDValue: String, newLoginIDValue: String) // update the existing value to the new value.
removeLoginID(loginIDKey: String, loginIDValue: String) // remove a login ID from the key's array, and remove the login ID Key itself if the array is empty
```

Haven't give any in-depth thought on this, please feel free to revise/propose an alternative based on real use cases.

Reply:
> I think in this design the Login ID Key is whitelist by a server-side configuration only so that's a non-issue? (In fact at #294 there is a option to whitelist it already)

Yes, I just explain why I opted a server-side config rather than doing it client-side.

> Can you be specific on why updateLoginID would be dangerous even for Client side with a logged in user session? Beware it is an API to update the LoginID, not other data of the principals.

Ideally, I'd like an authn system where if I hold a reference to a principal of user (i.e. principal ID), I can rely on it always referring the the same principal (i.e. immutable identity). However if we allow user to update login ID (i.e. changing the identity of an existing principal), this guarantee will not hold.

Reply:
> Yes, I just explain why I opted a server-side config rather than doing it client-side.

ic, but it comes out of nowhere, please start a different thread if it is a new discussion. So the discussion is traceable.

> Ideally, I'd like an authn system where if I hold a reference to a principal of user (i.e. principal ID), I can rely on it always referring the the same principal (i.e. immutable identity). However if we allow user to update login ID (i.e. changing the identity of an existing principal), this guarantee will not hold.

ic, but we don't have principals now. Instead, shall we have some data about the current session in authn response? No matter it is by password or other providers.

Reply:
Actually this thread is getting too long, let me resolve it and start a new thread. thanks.